File: 0287.find_the_duplicate_number/solution.go
package leetcode0287

func findDuplicate(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	fast, slow := 0, 0
	fast = nums[nums[fast]]
	slow = nums[slow]
	for nums[fast] != nums[slow] {
		fast = nums[nums[fast]]
		slow = nums[slow]
	}
	fast = 0
	for fast != slow {
		fast = nums[fast]
		slow = nums[slow]
	}
	return fast
}

File: 0287.find_the_duplicate_number/solution.py
from typing import List


class Solution:
    def find_duplicate(self, nums: List[int]) -> int:
        """ 类型环形链表II """
        slow = fast = 0
        fast = nums[nums[fast]]
        slow = nums[slow]
        while nums[slow] != nums[fast]:
            fast = nums[nums[fast]]
            slow = nums[slow]
        fast = 0
        while fast != slow:
            slow = nums[slow]
            fast = nums[fast]
        return fast
